Refactor update/setup logic

sirmarcel opened the pull request:
I think it's now a bit clearer what `update` is supposed to do, and how it does it:

- `self.atoms` is a record of the last `Atoms` for which the `Calculator` has done anything (we always need to store a copy, since we don't know what other parts of `ase` are fiddling with the atoms, and we don't want to miss any changes)
- `ase` will give use a new `Atoms` object via `calculate` or `get_property`, we then need to figure out if (a) there are absolutely no changes, in which case we can just return from `self.results` if asked, or if (b) something has changed
- if something has changed, we surely need to recompute `self.results` and in some cases, we need to do housekeeping (resetting neighbour lists, rebuilding displacements)

I also cleared up the use of `on_atoms_changed` to reset the `neighborlist` in the LJ calculator, I hope this is correct. The original version, if I understood it right, would have reset the `neighborlist` on *every* change of `self.atoms`, even if not needed.

sirmarcel committed:
Lightly refactor update/setup logic and state management and add rudimentary tests.

This should hopefully address some confusion caused by #4. Let's discuss!

fabiannagel committed:
Failing: on_atoms_changed called multiple times for changes in PBCs

fabiannagel commented on asax/calculator.py:
```diff
@@ -29,46 +29,34 @@ def update(self, atoms: Atoms):
+            if "cell" in changes or "pbc" in changes:
+                # need to recreate displacement; might as well start over
+                self.atoms = None
+                self.update(atoms)
```
Isn't there a `return` missing here? The current implementation causes two calls of `self.on_atoms_changed()` for cell or PBC changes. See the following commit with a failing test case for this.

sirmarcel replied:
> Isn't there a `return` missing here? The current implementation causes two calls of `self.on_atoms_changed()` for cell or PBC changes. See the following commit with a failing test case for this.

Hm, I don't understand what's going on there. I thought the call to `self.update()` would stop execution here and start from the top...? Strange.


> (Running the test cases individually doesn't expose this error, which is usually a hint that the unit tests are not atomic. I'd move the calculator initializations to `setUp()` to prevent stateful behavior here).

Why are there two tests for LJ anyway? It makes more sense in the `setUp()` for sure, that's what it's for... did I originally write this? If yes, shame on me.

fabiannagel committed:
Failing: Changes in cell and atomic positions breaks neighbor list update

fabiannagel commented on asax/lj.py:
```diff
@@ -41,16 +41,14 @@ def __init__(
         # non-PBC atom passed - disable stress computation
-        if not all(self.atoms.get_pbc()) and self.stress:
+        if not all(atoms.get_pbc()) and self.stress:
             self.stress = False
```
We probably also want to enable `stress` in the opposite case.

sirmarcel replied:
I think we should simply drop support for this, let it fail. I don't think people switching from molecule to material or vice versa with one calculator is an edge case, we can always support it if someone needs it. We certainly don't.

fabiannagel commented on asax/calculator.py:
```diff
@@ -29,46 +29,34 @@ def update(self, atoms: Atoms):
             self.results = {}
-            self.on_atoms_changed()
             self.setup()
             return
```
I'd put everything from line 32 to 37 into a dedicated method `reinitialize()` or something. IMHO this all is more confusing than it should be. `update()` should only check for changes in `atoms` and delegate accordingly. 

Setting `self.atoms = None` and recursively calling `self.update(atoms)` is essentially a complicated way to get to line 32 (what could be a simple call of `reinitialize()`).

sirmarcel replied:
Yes, that is correct. I'll do this. `update` in general is a bad name, but I can't think of anything better.

sirmarcel replied:
Ok, made a commit in that direction.

sirmarcel committed:
Further paring down of update

This removes any stateful behaviour from the sub-classes
of Calculator: the only interaction it has with the bigger
machinery of asax is then setup() is called, which happens
every time we detect a change of the structure that goes
beyond just positions wiggling around.

I think soon, we will need to implement a base class for
neighborlist-based potentials to deal with overflows and
stuff like that...

fabiannagel committed:
Logic moved to Calculator class

fabiannagel committed:
Atomic coordinates and box as jnp arrays
